modules/internal/ChapelIO.chpl

```diff
@@ -17,36 +17,310 @@
+ * :proc:`Reader.readWriteNewline` :proc:`Writer.readWriteNewline` reads and discards input until a newline or writes a newline.
+
+Objects of type :class:`Reader` also supports the following methods:
+
+ * :proc:`Reader.read` (similar to :proc:`IO.channel.read`)
+ * :proc:`Reader.readln` (similar to :proc:`IO.channel.readln`)
+
+Objects of type :class:`Writer` also support the following methods:
+
+ * :proc:`Writer.write` (similar to :proc:`IO.channel.read`)
+ * :proc:`Writer.writeln` (similar to :proc:`IO.channel.writeln`)
+
+Note that objects of type :class:`Reader` or :class:`Writer` may represent a
+locked channel; as a result, using parallelism constructs to call methods on
+:class:`Reader` or :class:`Writer` may result in undefined behavior.
+
+.. note::
+
+  In the future, we plan to merge :class:`Reader`, :class:`Writer`, and
+  :record:`IO.channel` in order to support calling :proc:`IO.channel.readf`
+  and other I/O routines inside a ``readThis`` method.
+
+Because it is often more convenient to use an operator for I/O, instead of
+writing
+
+.. code-block:: chapel
+
+  f.readwrite(x);
+  f.readwrite(y);
+
+one can write
+
+.. code-block:: chapel
+
+  f <~> x <~> y;
+
+Note that the types :type:`IO.ioLiteral` and :type:`IO.ioNewline` may be useful
+when using the ``<~>`` operator. :type:`IO.ioLiteral` represents some string
+that must be read or written as-is (e.g. ``","`` when working with a tuple),
+and :type:`IO.ioNewline` will emit a newline when writing but skip to and
+consume a newline when reading.
+
+
+This example defines a readWriteThis method and demonstrates how ``<~>`` will
+call the read or write routine, depending on the situation.
+
+.. code-block:: chapel
+
+  class IntPair {
+    var x: int;
+    var y: int;
+    proc readWriteThis(f) {
+      f <~> x <~> new ioLiteral(",") <~> y <~> new ioNewline();
+    }
+  }
+  var ip = new IntPair(17,2);
+  write(ip);
+  // prints out
+  // 17,2
+
+  delete ip;
+
+This example defines a only a writeThis method - so that there will be a
+function resolution error if the class NoRead is read.
+
+.. code-block:: chapel
+
+  class NoRead {
+    var x: int;
+    var y: int;
+    proc writeThis(f:Writer) {
+      f.writeln("hello");
+    }
+    // Note that no readThis function will be generated.
+  }
+  var nr = new NoRead();
+  write(nr);
+  // prints out
+  // hello
+
+  // Note that read(nr) will generate a compiler error.
+
+  delete nr;
+
+
+Generalized write and writeln
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The :class:`Writer` class contains no arguments and serves as a base class to
+allow user-defined classes to be written to.  If a class is defined to be a
+subclass of Writer, it must override the :class:`Writer.writePrimitive` method
+that any Chapel primitive type as an argument.
+
+The following code defines a subclass of :class:`Writer` that overrides the
+:proc:`Writer.writePrimitive` method to allow it to be written to.  It also
```

Comment:
Since they've seen the prefix in the last reference to writePrimitive, maybe we should shave it off of this reference here?